### PR TITLE
Use jest

### PR DESCRIPTION
**Pull request by danez**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | #6179 #6442
| Patch: Bug Fix?          |n
| Major: Breaking Change?  | n
| Minor: New Feature?      | n
| Tests Added + Pass?      | Yes
| Documentation PR         | 
| Any Dependency Changes?  |
| License                  | MIT

Rebase of #6179. Seems i removed my fork repo. :)

All tests work, babel-register is skipped for now. But I have an idea how to fix it, by mocking pirates and just checking what babel-register is doing, pirates is covered with tests anyway so we don't have to test it.

### CODE REVIEW

File: test/testSetupFile.js [hunk 1]
```diff
@@ -0,0 +1,2 @@
+/* eslint-env jasmine */
+jasmine.DEFAULT_TIMEOUT_INTERVAL = 10000;
```
Comment (SimenB) [inline test/testSetupFile.js]: you can use [`jest.setTimeout`](https://facebook.github.io/jest/docs/en/jest-object.html#jestsettimeouttimeout)

File: .babelrc.js [hunk 2]
```diff
@@ -32,11 +43,4 @@ if (process.env.BABEL_ENV === "cov") {
   config.plugins.push("babel-plugin-istanbul");
```
Comment (SimenB) [inline .babelrc.js]: should this still be here? Jest adds it itself (and you removed it from package.json)

Comment (danez) [reply .babelrc.js]: good catch

File: test/testSetupFile.js (hunk 1 above)
Comment (danez) [reply test/testSetupFile.js]: Fixed

File: .babelrc.js [hunk 3]
```diff
@@ -40,7 +40,6 @@ const config = {
 
 if (process.env.BABEL_ENV === "cov") {
   config.auxiliaryCommentBefore = "istanbul ignore next";
```
Comment (SimenB) [inline .babelrc.js]: jest also does this

Comment (danez) [reply .babelrc.js]: :) even better

Comment (SimenB) [reply .babelrc.js]: https://github.com/facebook/jest/blob/a88d3c553d94b220f9dae47f64983d9d7b3baed8/packages/jest-runtime/src/script_transformer.js#L153-L169

**Comment by hzoo** [comment]
@danez maybe you didn't see https://github.com/babel/babel/pull/6442? we were going to merge that but not sure the relation @rogeliog

**Comment by danez** [comment]
I think there is no need to use the mocha runner, because just using the default jest runner works fine.

File: .travis.yml [hunk 4]
```diff
@@ -7,9 +7,9 @@ cache:
   directories:
   - node_modules
 node_js:
+  - '9'
   - '8'
   - '6'
-  - '4'
```
Comment (hzoo) [inline .travis.yml]: This seems unrelated? We are we dropping node 4 testing already?

File: packages/babel-code-frame/src/index.js [hunk 5]
```diff
@@ -180,15 +180,17 @@ export default function(
   if (!deprecationWarningShown) {
     deprecationWarningShown = true;
 
-    const deprecationError = new Error(
```
Comment (hzoo) [inline packages/babel-code-frame/src/index.js]: I'm ok with just not doing the warning anymore and making this a breaking change. We can just fix this automatically with https://github.com/babel/babel-upgrade

File: .travis.yml (hunk 4 above)
Comment (existentialism) [reply .travis.yml]: Can use `node` environment and keep 4? (like in babel-upgrade)

Comment (hzoo) [reply .travis.yml]: If it's because of jest, it still works on node 4 if you use the node environment - https://github.com/babel/babel-upgrade works fine with it.

Comment (danez) [reply .travis.yml]: ok cool, I revert then.

File: packages/babel-code-frame/src/index.js (hunk 5 above)
Comment (danez) [reply packages/babel-code-frame/src/index.js]: We can do this in a separate PR.

**Comment by rogeliog** [comment]
I agree with @danez! The Mocha runner is a good option when the complexity of the codebase makes it hard to fully migrate to Jest or as an intermediate step. Seem that I'll close it in favor of this, I can reopen it if needed.

**Comment by hzoo** [comment]
Ok well if this works I'm for it too 👍 , thanks for your efforts @rogeliog, sorry we didn't actually end up merging the PR 😅

**Comment by danez** [comment]
Ok I hope the coverage is now correct. I had to readd the babel-plugin-istanbul as some of our tests test the code in `lib` and not `src`. Wherever possible I changed to test `src` directly, but the fixture runner which runs stuff in separate envs couldn't be changed easily.

**Comment by SimenB** [comment]
Anything jest can do to help, or is it just a quirk (feature) of babel's tests?

**Comment by danez** [comment]
I think that we in the past implemented some features that jest might already do, like running each test in a spearate vm (https://github.com/babel/babel/blob/master/packages/babel-helper-transform-fixture-test-runner/src/index.js#L87). 
The other thing is that we bundle some packages with rollup/webpack and we want to test the bundle rather than the source in this cases to ensure that these works.
We might be able to get rid of some of this now, but I want to make this PR as small as possible.

**Comment by Andarist** [comment]
Super cool 👍I was wondering - it should be possible to migrate to builtin snapshots instead of custom setup in the future, right? This would allow easier test updates

**Comment by hzoo** [comment]
I actually like the individual files lol compared to reading the snapshot output now lol

**Comment by Andarist** [comment]
What's the difference though? A single snapshot format? Maybe it's possible to integrate with snapshot system to output whatever we want in whatever directory we want, just have it backed by jest's built-in system for this. `expected` files are a form of snapshots anyway

**Comment by hzoo** [comment]
Oh ok if you can change the format then that's cool. I just mean https://github.com/babel/babel-upgrade/blob/51534a4bd3f742d284d42e231ae36e71c83bc93d/__tests__/__snapshots__/packageData.test.js.snap#L5-L12 vs the actual js

**Comment by Andarist** [comment]
Actually I have no idea if its possible to change the whole format, it's for sure possible to write custom serializer, but not sure if a snapshot file can be an actual js file, not assigned to exports etc (jest snapshots are actually requireable). Worth investigating in the future
